Implement WGSL abstract types for global `const` declarations and constructor calls.

jimblandy opened the pull request:
**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `cargo clippy`.
- [x] Run `cargo xtask test` to run tests.
- [x] Add change to `CHANGELOG.md`. See simple instructions inside file.

jimblandy committed:
[naga] Introduce `ScalarKind::AbstractInt` and `AbstractFloat`.

Introduce new variants of `naga::ScalarKind`, `AbstractInt` and
`AbstractFloat`, for representing WGSL abstract types.

jimblandy committed:
[naga] Introduce `Literal::AbstractInt` and `AbstractFloat`.

Introduce new variants of `naga::Literal`, `AbstractInt` and
`AbstractFloat`, for representing WGSL abstract values.

jimblandy commented:
This is still dependent on a bunch of other PRs, visible in the merge history. Once those are landed, and I've addressed CI issues, I'll take this out of draft.

jimblandy commented:
Regarding the mess described in #4747: this PR just deletes `tests/in/glsl/double-math-functions.frag` for the time being, as it can't be translated to WGSL.

jimblandy committed:
[naga wgsl-in] Reformat match statement for better patch stability.

The large `match` statement in `Lowerer::construct` seems to flop back
and forth between two indentation levels as it's edited, making the
diffs hard to read. Rewrite it to use deferred initialization of
`expr`, so that `cargo fmt` doesn't have to decide whether or not to
put the `match` on the same line as `let expr`. This makes subsequent
diffs easier to read.

jimblandy committed:
[wgsl-in]: Remove `Components::Many::first_component_ty_inner`.

Delete the `first_component_ty_inner` field from
`front::wgsl::lower::construction::Components::Many`. With the
introduction of abstract types, it will no longer be possible to infer
the type of the vector being constructed by looking at the type of its
first constructor argument alone: automatic conversion rules might
need to be applied to that argument.

jimblandy committed:
[naga wgsl-in] Implement abstract types for consts, constructors.

jimblandy commented:
Since #4747 has landed, this is ready for review.

jimblandy commented:
By the way - I think we should sometime soon rename `Lowerer::expression` to `concretized_expression`, and then rename `expression_for_abstract` to `expression`. Since so many operations in WGSL accept abstract types, the "natural" choice should the one that doesn't concretize yet, and code that does concretize should look "suspicious", as potentially in need of an update.

(In #4755 I was confused for a bit by test failures before I realized that I was still automatically concretizing the initialization expression, so that coercion to the var's declared type didn't work.)

But that can be a separate PR, isolated from any semantic changes.

jimblandy commented:
Hmm. There is a slightly different way to approach this patch, which would be to make constant evaluation responsible for detecting when conversions are necessary and applying them. The WGSL front end is still going to need to do things like choose types for all those `Compose` expressions in `construction.rs`, for example, but once it's done so, it seems like the actual conversions could be done directly in `constant_evaluator.rs`, rather than having the front end create a bunch of `As` expressions which it trusts the constant evaluator to consume.

@teoxoy If you'd like me to try out that alternative, just let me know. I don't think it'd take long.

teoxoy commented on naga/src/valid/expression.rs:
```diff
@@ -670,22 +670,36 @@ impl super::Validator {
                     Bo::Add | Bo::Subtract => match *left_inner {
                         Ti::Scalar(scalar) | Ti::Vector { scalar, .. } => match scalar.kind {
-                            Sk::Uint | Sk::Sint | Sk::Float => left_inner == right_inner,
+                            Sk::Uint
+                            | Sk::Sint
+                            | Sk::Float
+                            | Sk::AbstractInt
+                            | Sk::AbstractFloat => left_inner == right_inner,
                             Sk::Bool => false,
```
Shouldn't we return `false` for `AbstractInt`/`AbstractFloat` (in all the other places in this file as well)? It would make things clearer when reading the code.

jimblandy replied:
You're right. Changed as suggested.

teoxoy commented on naga/src/front/wgsl/lower/conversion.rs:
```diff
@@ -0,0 +1,378 @@
+        Some((expr_scalar, goal_scalar))
+    }
+
+    fn automatically_convertable_scalar(
```
```suggestion
    fn automatically_convertible_scalar(
```

jimblandy replied:
... I have no excuse ...

teoxoy commented on naga/src/front/wgsl/lower/conversion.rs:
```diff
@@ -0,0 +1,378 @@
+                kind: goal_scalar.kind,
+                convert: Some(goal_scalar.width),
+            };
+            log::trace!("JIMB: Emitting {cast:?}");
```
```suggestion
```

jimblandy replied:
I've removed most of these, and cleaned up the one I wanted to keep.

teoxoy commented on naga/src/front/wgsl/lower/conversion.rs:
```diff
@@ -0,0 +1,378 @@
+        use crate::ScalarKind as Sk;
+        use crate::TypeInner as Ti;
+
+        log::trace!("JIMB: automatically_converts_to: {self:?} -> {goal:?}");
```
```suggestion
```

teoxoy commented on naga/src/front/wgsl/lower/construction.rs:
```diff
@@ -546,3 +591,45 @@ impl<'source, 'temp> Lowerer<'source, 'temp> {
+    components: &[Handle<crate::Expression>],
+    ctx: &ExpressionContext<'_, '_, '_>,
+) -> Result<Scalar, usize> {
+    log::trace!("JIMB: automatic_conversion_consensus");
```
```suggestion
```

teoxoy commented on naga/src/front/wgsl/lower/conversion.rs:
```diff
@@ -0,0 +1,378 @@
+            _ => return None,
+        }
+
+        log::trace!("      okay: expr {expr_scalar:?}, goal {goal_scalar:?}");
```
```suggestion
```

teoxoy commented on naga/src/front/wgsl/lower/construction.rs:
```diff
@@ -546,3 +591,45 @@ impl<'source, 'temp> Lowerer<'source, 'temp> {
+        .iter()
+        .map(|&c| ctx.typifier()[c].inner_with(types));
+    let mut best = inners.next().unwrap().scalar().ok_or(0_usize)?;
+    log::trace!("      start: {best:?}");
```
```suggestion
```

teoxoy commented on naga/src/front/wgsl/lower/construction.rs:
```diff
@@ -546,3 +591,45 @@ impl<'source, 'temp> Lowerer<'source, 'temp> {
+        match best.automatic_conversion_join(scalar) {
+            Some(new_best) => {
+                best = new_best;
+                log::trace!("      new:   {best:?}");
```
```suggestion
```

teoxoy commented on naga/src/front/wgsl/lower/conversion.rs:
```diff
@@ -0,0 +1,378 @@
+            }
+
+            // AbstractInt converts to AbstractFloat.
+            (Sk::AbstractFloat | Sk::AbstractInt, Sk::AbstractFloat | Sk::AbstractInt) => {
+                Some(Self {
+                    kind: Sk::AbstractFloat,
+                    width: crate::ABSTRACT_WIDTH,
+                })
+            }
+
+            // AbstractFloat converts to Float.
+            (Sk::AbstractFloat, Sk::Float) => Some(Self::float(other.width)),
+            (Sk::Float, Sk::AbstractFloat) => Some(Self::float(self.width)),
+
+            // AbstractInt converts to concrete integer or float.
+            (Sk::AbstractInt, kind @ (Sk::Uint | Sk::Sint | Sk::Float)) => Some(Self {
+                kind,
+                width: other.width,
+            }),
+            (kind @ (Sk::Uint | Sk::Sint | Sk::Float), Sk::AbstractInt) => Some(Self {
+                kind,
+                width: self.width,
+            }),
```
```suggestion
            (Sk::AbstractFloat, Sk::AbstractInt) => Some(other),
            (Sk::AbstractInt, Sk::AbstractFloat) => Some(self),

            // AbstractFloat converts to Float.
            (Sk::AbstractFloat, Sk::Float) => Some(other),
            (Sk::Float, Sk::AbstractFloat) => Some(self),

            // AbstractInt converts to concrete integer or float.
            (Sk::AbstractInt, Sk::Uint | Sk::Sint | Sk::Float) => Some(other),
            (Sk::Uint | Sk::Sint | Sk::Float, Sk::AbstractInt) => Some(self),
```

I find this easier to follow.

jimblandy replied:
Very much so.

teoxoy commented on naga/src/front/wgsl/lower/conversion.rs:
```diff
@@ -0,0 +1,378 @@
+    /// If there are any scalars to which WGSL's automatic conversions
+    /// will convert both `self` and `other`, return the best such
+    /// scalar. Otherwise, return `None`.
+    pub const fn automatic_conversion_join(self, other: Self) -> Option<crate::Scalar> {
```
```suggestion
    pub const fn best_common_conversion(self, other: Self) -> Option<crate::Scalar> {
```

Is this better?

jimblandy replied:
I want to have the phrase "automatic conversion" in there, because this is supposed to be limited the WGSL "automatic conversions".

jimblandy replied:
But I agree that "join" is a bit jargony.

teoxoy reviewed (commented):
Solid stuff!

Left a few suggestions.

teoxoy commented:
> Hmm. There is a slightly different way to approach this patch, which would be to make constant evaluation responsible for detecting when conversions are necessary and applying them. The WGSL front end is still going to need to do things like choose types for all those `Compose` expressions in `construction.rs`, for example, but once it's done so, it seems like the actual conversions could be done directly in `constant_evaluator.rs`, rather than having the front end create a bunch of `As` expressions which it trusts the constant evaluator to consume.
> 
> @teoxoy If you'd like me to try out that alternative, just let me know. I don't think it'd take long.

If it's possible that would be great as we are now injecting `As` expressions that just end up being evaluated by the constant evaluator later anyway.

jimblandy committed:
[naga wgsl-in] Clarify match in `automatic_conversion_join`.

Co-authored-by: Teodor Tanasoaia <28601907+teoxoy@users.noreply.github.com>

jimblandy committed:
[naga] Improve snapshot output when validation fails.

jimblandy committed:
[naga wgsl-in] Matrix constructors have only float overloads.

jimblandy committed:
[naga wgsl-in] Constructors with types don't make abstract values.

When a constructor builtin has an explicit type parameter, like
`mat2x2<f32>`, it should not produce an abstract matrix, even if its
arguments are abstract.

jimblandy commented:
@teoxoy Just added one more commit, nothing else changed.

teoxoy commented:
Looks good, let's land it.
